contracts/VetoGuard.sol

```diff
@@ -0,0 +1,237 @@
+    /// @param initializeParams Parameters of initialization encoded
+    function setUp(bytes memory initializeParams) public override initializer {
+        __Ownable_init();
+        (
+            uint256 _executionDelayBlocks,
+            address _owner,
+            address _vetoERC20Voting,
+            address _gnosisSafe
+        ) = abi.decode(initializeParams, (uint256, address, address, address));
+
+        executionDelayBlocks = _executionDelayBlocks;
+        transferOwnership(_owner);
+        vetoERC20Voting = IVetoERC20Voting(_vetoERC20Voting);
+        gnosisSafe = IGnosisSafe(_gnosisSafe);
+
+        emit VetoGuardSetup(
+            msg.sender,
+            _executionDelayBlocks,
+            _owner,
+            _vetoERC20Voting,
+            _gnosisSafe
+        );
+    }
+
+    /// @notice Allows a user to queue the transaction, requires valid signatures
+    /// @param to Destination address.
+    /// @param value Ether value.
+    /// @param data Data payload.
+    /// @param operation Operation type.
+    /// @param safeTxGas Gas that should be used for the safe transaction.
+    /// @param baseGas Gas costs for that are independent of the transaction execution(e.g. base transaction fee, signature check, payment of the refund)
+    /// @param gasPrice Maximum gas price that should be used for this transaction.
+    /// @param gasToken Token address (or 0 if ETH) that is used for the payment.
+    /// @param refundReceiver Address of receiver of gas payment (or 0 if tx.origin).
+    /// @param signatures Packed signature data ({bytes32 r}{bytes32 s}{uint8 v})
+    function queueTransaction(
+        address to,
+        uint256 value,
+        bytes memory data,
+        Enum.Operation operation,
+        uint256 safeTxGas,
+        uint256 baseGas,
+        uint256 gasPrice,
+        address gasToken,
+        address payable refundReceiver,
+        bytes memory signatures
+    ) external {
+        bytes32 transactionHash = getTransactionHash(
+            to,
+            value,
+            data,
+            operation,
+            safeTxGas,
+            baseGas,
+            gasPrice,
+            gasToken,
+            refundReceiver
+        );
+
+        require(
+            transactionQueuedBlock[transactionHash] == 0,
+            "Transaction has already been queued"
+        );
+
+        bytes memory gnosisTransactionHash = gnosisSafe.encodeTransactionData(
+            to,
+            value,
+            data,
+            operation,
+            safeTxGas,
+            baseGas,
+            gasPrice,
+            gasToken,
+            refundReceiver,
+            gnosisSafe.nonce()
+        );
+
+        // If signatures are not valid, this will revert
+        gnosisSafe.checkSignatures(
+            keccak256(gnosisTransactionHash),
+            gnosisTransactionHash,
+            signatures
+        );
+
+        transactionQueuedBlock[transactionHash] = block.number;
```

Comment:
Not entirely sure about it but is it possible this could be frontrun, for example by buying up  tokens if it's seen that a new transaction is being queued?

In which case `block.number - 1;` might be better here?

Reply:
@tbwebb22 @christopherdancy is @herbig's comment a valid concern?

Reply:
Users would already be able to see a transaction that is getting signed well before it gets queued, so I think the risk here is pretty low.

However, I do think using the `block.number - 1` is a good practice to prevent front running regardless.  

This makes more sense to me to implement in the veto voting contract, so I pushed up a change just now to count the votes from 1 block before the transaction gets queued:
![image](https://user-images.githubusercontent.com/84364476/191825719-a75ce9a4-462c-46ef-8234-4a4abd681aef.png)

Reply:
yeah makes sense.  I didn't realize they could see it before it's queued

Reply:
Yeah, there won't be any on-chain indication typically of the tx, but users will be able to see the tx on our UI before its queued, as people begin signing the tx. Before it gets queued / executed, the tx really only exists off-chain in the Gnosis API.

Reply:
I suppose in that sense anyone can prepare for a potential vote and acquire tokens if they feel that strongly about something.